### ISSUE TITLE
Support complex dtype for jax.random.truncated_normal

### ISSUE THREAD

**Issue by sracaniere** [google/jax#8448]
`jax.random.normal` supports both float and complex dtypes. It handles complex dtypes by treating C^n as if it was R^{2n}. Things are different for `jax.random.truncated_normal` where only float types are supported. Could we have the same approach for `jax.random.truncated_normal` and `jax.random.normal`, and treat C^n as R^{2n}? This would make it easier to use complex dtypes.

As a mathematician myself, I find it quite natural to treat C^n as R^{2n}, and this approach is consistent with the current behaviour of `jax.random.normal`.

**Comment by jakevdp** [comment]
One point of clarification: what convention should we use to truncate the complex normal? Are we truncating the magnitude of the complex outputs (i.e. creating an annulus in thecomplex plane) or are we truncating the components (i.e. creating a rectangle in the complex plane)?

**Comment by sracaniere** [comment]
In R^n, the truncation is done inside a box, not inside a region bounded by two spheres. If we decide to see the complex plane as R^2, then we conclude truncation should indeed be done in a rectangle. The same conclusion carries over to C^n since we use C^n = R^{2n} as vector spaces over R, and we should truncate inside a rectangle.

**Comment by jakevdp** [comment]
OK, that makes sense. What API do you have in mind for specifying the lower and upper bounds of the complex & imaginary parts? Perhaps if the bounds are complex, the imaginary part should give the imaginary bound, and the real part should give the real bound?

**Comment by jakevdp** [comment]
Also, mathematically is it true that sampling x and y independently from truncated normal distributions is equivalent to sampling from a 2D truncated normal within a rectangle? It's not obvious to me that this would be the case.

**Comment by sracaniere** [comment]
Regarding the API. What you suggest sounds reasonable to me. So, if the dtype is complex we could have the following cases:
1. upper and lower are a single float. This should specify the upper and lower bounds of both the real and imaginary
2. upper and lower are a single complex. The real parts specify the lower & upper bounds of the real output, the imaginary parts of the upper/lower specify the lower/upper parts of the outputs

The above are for the case where the inputs are a single number. If the inputs are arrays, we could also have two cases, or require that the dtype be complex. I can see arguments for both (flexibility vs simplicity). What do you think?

**Comment by sracaniere** [comment]
Regarding your question "is it true that sampling x and y independently from truncated normal distributions is equivalent to sampling from a 2D truncated normal within a rectangle?", I think it's true, but I'll have to check and answer tomorrow. It's late here in London, and I've just been told off by my wife for working late. lol

**Comment by PhilipVinc** [comment]
Just FYI: there's already an implementation of `truncated_normal` (and of `truncated_uniform`) living in the initialisers, though it truncates the magnitude.
The need for this kind of truncation arises from the paper linked in https://github.com/google/jax/pull/7670 (having glory & co initialisers work with complex valued layers).

Almost one year ago @jekbradbury [had expressed concerns](https://github.com/google/jax/issues/5312#issuecomment-755032202) whever there is consensus on the definition of those distributions.

**Comment by sracaniere** [comment]
@PhilipVinc Thanks for the pointers. Reading your comments there, I can indeed see a lack of consensus. My suggestion was to treat C as R^2 everywhere. This is not a surprising choice, and it would give a consistent choice of how to extend many random variables from R^n to C^n. Another advantage of this approach is simplicity of the code: sampling in C^n is just sampling in R^n. But it will not help if what you need is a truncated normal with a center and a radius in C.

@jakevdp What do you think? My two cents: if we decide against adding support for C, we should then update the doc of the truncated normal briefly explaining why there is no support for complex dtype. This will prevent end users from complaining about it, and make them understand they need to implement this themselves depending on their preferred method.

**Comment by sracaniere** [comment]
@jakevdp Regarding your question "is it true that sampling x and y independently from truncated normal distributions is equivalent to sampling from a 2D truncated normal within a rectangle?".

Let p(x) denote the density of a random variable on R. Use the same letter p(x, y) when sampling two independent variables x and y from the same previous distribution, so that that p(x, y) = p(x)p(y). In our case of interest, p is the density of the normal (not truncated) distribution on R. I think your question is equivalent to asking, "is it true that `p(x, y| a<x<b, a<y<b) = p(x| a<x<b)p(y| a<y<b)` ?". Let's give these two densities names: `h(x, y) = p(x, y| a<x<b, a<y<b)` and `g(x, y) = p(x| a<x<b)p(y| a<y<b)`.
Let's compute them. First of all, both h and g are 0 outside of a rectangle. Inside that rectangle: `h(x, y) = p(x, y) / Z`, where Z is some normalising constant, and `g(x, y) = p(x)p(y) / (Z_x Z_y)` where `Z_x` and `Z_y` are also normalising constants. Let's compute from the definition of Z:
```Z = \int_{a<x<b, a<y<b} dx dy p(x, y) = \int_{a<x<b, a<y<b} dx dy p(x)p(y) = \int_{a<x<b} dx dy p(x) \int_{a<y<b} dx dy p(y).```
This last term is exactly the product `Z_x Z_y`. This proves that indeed sampling x and y independently from truncated normal distributions is equivalent to sampling from a 2D truncated normal within a rectangle. This statement is also true for other distributions, since I did not use the fact that `p` was a normal.

**Comment by jakevdp** [comment]
Thanks, that makes sense.

So I think we're left in the same place as previously: it's difficult to agree on whether distributions of complex numbers should be defined along cartesian or radial coordinates in the complex space... the non-truncated gaussian has the happy coincidence that both approaches are equivalent, so it made sense to implement it.

But it's not clear to me what the best default is for other distributions.

**Comment by jakevdp** [comment]
I'm going to close this for a similar reason to that in #4680: it's not clear what API we should choose for complex-valued random number generators, so it's best to leave that up to the user.
